### PR TITLE
Added Prettier.io code formatter

### PR DESCRIPTION
**Pull request by branneman**
This PR adds [Prettier](https://prettier.io/), changes [ESLint](https://eslint.org/) config to just the functional rules, and removes [stylelint](https://stylelint.io/). Prettier supports auto-formatting MD, CSS, SCSS, JS and JSON.

This PR and Prettier come with a defined, opinionated codestyle. Adopting this standard style means ranking the importance of code clarity and community conventions higher than personal style. The whole point of this PR is to save you time by avoiding [bikeshedding](https://www.freebsd.org/doc/en/books/faq/misc.html#bikeshed-painting) about code style. There are lots of debates online about tabs vs. spaces, etc. that will never be resolved. These debates just distract from getting stuff done. At the end of the day you have to 'just pick something', and that's what this PR does: its a bunch of sensible 'just pick something' opinions. Hopefully, users see the value in that over defending their own opinions. There are actual real problems that you could spend your time solving!

Two new npm scripts are made available in this PR:
- `npm run codestyle`: This script automatically formats all supported languages with Prettier
- `npm run codestyle-is-valid`: This script checks if the code style passes, and exits with a non-zero code if it does not pass.

Prettier allows for some style rules, located in the `prettier` property in the `package.json`, however I advice strongly against changing any rules, because that defeats the purpose.

Some related additional changes:
- Moved from json to yaml for the ESLint config file: `.eslintrc.yml`, so comments are supported.
- Updated ESLint version (via `babel-eslint` and `gulp-eslint`).
- Updated githooks: `pre-commit` now also runs `npm run codestyle-is-valid` when a match for `*.{md,css,scss,js,json}` was found, blocking the commit when the code does not pass the style. It also still blocks the commit when linting fails.
- Appended `.min` to the filenames for minified files, so they are ignored by prettier and linting: `bootstrap.min.css` and `highlight.min.js`.

I did not squash the 2 commits because otherwise the actual functional changes will get mingled up together with the huge amount of code formatting changes. See commit [423bc3d](https://github.com/mirabeau-nl/frontend-bootstrap/commit/423bc3d) for just the new features.

### CODE REVIEW

**Comment by waarissyb** [comment]
Nice job :) I’ve got some feedback about feedback.

When running `npm run codestyle-is-valid` in a codebase with some bad formatted code, the following is returned:
```
[…]
npm ERR! Failed at the frontend-bootstrap@1.0.0 codestyle-is-valid script 'prettier --list-different --ignore-path .eslintignore "**/*.{md,css,scss,js,json}"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the frontend-bootstrap package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     prettier --list-different --ignore-path .eslintignore "**/*.{md,css,scss,js,json}"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs frontend-bootstrap
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls frontend-bootstrap
npm ERR! There is likely additional logging output above.
[…]
```
As you described in your PR, this is the correct output. I end up with a non-zero code. However, it isn’t clear feedback to the user. It seems like something went wrong. When running `gulp lint` I end up with something like this:
```
✖ 2579 problems (2562 errors, 17 warnings)
```
That’s better feedback. Is it possible to let the script output something similar?

I ran `npm run codestyle` successfully. If I understood correctly, the files that were OK are grey in the output. Files that are corrected are placed white. It isn’t that clear. Maybe it’s better to end with a summary of what the script has done, like with `gulp lint`.

When running `npm run codestyle-is-valid` with correct code, I get no positive feedback :(

I see that you’ve changed the indenting format from 4 to 2 spaces. I’m fine with that, but then you should reformat the HTML (njk-files) as well in this PR.

**Comment by branneman** [comment]
Thanks for the feedback, will update PR. Not sure how far we should go on getting nicer error messages, but it's a good suggestion.

It also turns out there are non-stylistic rules in Stylelint, namely [Possible errors](https://stylelint.io/user-guide/rules/#possible-errors) and [Limit language features](https://stylelint.io/user-guide/rules/#limit-language-features).

I will re-add stylelint to the PR as well with just those rules, in line with ESLint.

**Comment by awolkers** [comment]
Can you re investigate the way ho the eslint rules are added. In the old situation we start with the recommended settings from eslint and overwrite those with our own.
`"extends": "eslint:recommended"`

Prettier also defines a settings file which we can include bij default . Check for more info [Prettier ESlint integration](https://prettier.io/docs/en/eslint.html)

> The easiest way to do this is to use [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier). It's a one liner that can be added on-top of any existing ESLint configuration.

We might want to use something like:
```
{
  "extends": [
    "eslint:recommended",
    "prettier"
  ]
},
// our own rules
```

**Comment by MartijnNieuwenhuizen** [comment]
Let me start by saying a generic code style is something I really like. But of course, I would prefer it to be the style I'm currently using. Therefor I tool a look at what would change for me. A lot of those things are are ignorable, but there are things I have some difficulty with:

## 1. Indentation
As a person with Dyslexia, while reading you will often experience the fact that you accidentally will jump to another line. A larger indentation (4 spaces) increases this. Therefor I believe the change of indentation from 4 to 2 spaces reduces the ability to scan the code more quickly.

### 1.1 Indentation in .njk files
There is also a problem here with the indentation of the .njk files. If you use a component from another project, or pull this update in your current project, Prettier will fix the indentations of the .js and .scss files. However, the .njk files will remain with a indentation of 4 spaces. If you start adding new lines to it, these are indented with 2 spaces. I think this should be something we would like to be fixed as well.

## 2. Semicolons
Also the use of semicolons will help you spot the end of the line more easily. This will avoid you to expediently jump to anther line while reading and I believe it makes the code is easier to read if you include them. 

If we do include Prettier I'm wondering if there is a way to set the non-use of semicolons as a setting in Editorconfig as well.

## 3. The time it takes for the code to change
The fact that this tool changes your code is something I still find a bit scary. But that's not the part that's stops me. The amount of time that is takes to change the code does. 

I found it useful to run the prettier script the moment I finished a block of code. That's also the moment you move to the browser and take a look if your changes worked. Because Prettier takes some time to analyse and change your code, the moment that the changes a implemented is the moment you are in your browser for a second or 2. A change in code also makes your browser reload and if you are working on a part where the user input is handled by JS, it happens that you will have to tab a button twice if you for instance work on a popup or accordion. This is something I could imagine that would be irritating after a view times.

## CSS 
The way prettier changes your .js files is working smoothly. However there are some things I would like to do in the CSS files as well.

### Order in CSS
If it's possible to stick to the order of specificity, I would like it to reorder the properties as well. Here I mean that this:

```
.element {
  top: 0;
  display: block;
  color: lime;
  z-index: 10;
  position: absolute;
  background-color: red;
}
```

Becomes this:

```
.element {
  position: absolute;
  top: 0;
  display: block;
  z-index: 10;
  background-color: red;
  color: lime;
}
```

This would improve the order of CSS and therefor the speed you would be able to find a declaration.

### Fix shorthand
Sometimes I type:

```
margin: 0 auto 0 auto;
```

This could of course be written as

```
margin: 0 auto;
```

This isn't something that's fixed by Prettier and I think it should. And yes, the linter gives me a warning, but if it fixed so much of JS, this is something I would like as well.

## Learning
But there are also some things I like. And the biggest one is that you could learn from your from prettier. This is something I (re)learned:
```
return (a + b) === c
```
Could also be 
```
return a + b === c
```
This is of course a small thing, but I know there is more to discover!

## If we do it, we should do it right
Another thing I like is that it will avoid complex lines of code. In my current project this is something that happens sometimes, but that's mostly in the NodeJS part. And that brings me to the next subject. If we include this into our new projects I think we should also make the agreement that every project with a JS based backend should use Prettier as well. Otherwise a developer needs to switch code styles within a project and I think that's something that will really affect productivity.

## For the new ones
Oke, I'm nearly done here just the last thing. What about people who just start writing code (or start learning new things). I could foresee that it's more difficult to experiment because Prettier keeps changing your code and you won't immediately recognise where a part was. This isn't something I experience but something I foresee.

## Conclusion
I don't see enough benefits yet to include this right now. I'm also very curios about the opinions of the other colleagues. Therefor I would like to encourage a discussion about it, this could be here or on the next Guild day.

**Comment by nienkedekker** [comment]
I think Prettier is great and this is a good addition to the boilerplate. However, when I run `npm codestyle-is-valid` I get the same vague error Syb mentioned in https://github.com/mirabeau-nl/frontend-boilerplate/pull/76#issuecomment-354423199. Did you update the PR?

The only thing I'm unhappy with is forbidding the use of semi-colons. I'm a big proponent of them because it makes reading code so much easier: I can tell when a block ends. So I'm with Martijn on that one.

I also agree with Martijn re: CSS property ordering. Perhaps we could add something like this: https://github.com/hudochenkov/stylelint-order. I think the best way of ordering properties is  positioning, box model, typographic, and then visual. This may be out of scope for this PR though, but it's something we could talk about next Guild day.

TLDR: I like it but I need my semicolons 😅

**Comment by branneman** [comment]
@awolkers:
Good point, I will re-investigate and get back.

---
@MartijnNieuwenhuizen & @nienkedekker:
Semicolons: Talking about the use or absence of semicolons is bikeshedding by my definition. You will get used to either situation quicker than you expect (ask your colleagues!). I've chosen them because of StandardJS. Therefore I say your argument is invalid ;)

---
@MartijnNieuwenhuizen:
Thanks for your reply!

Indentation: It's interesting to hear that for somebody with Dyslexia a file that is indented with 2 spaces might be harder to read than that same file indented with 4 spaces. But are we sure about that? Is there any proof of that? Or is it temporary, and thus bikeshedding? I think readability can be increased by configuring your IDE with some friendlier typography settings: greater line-height, larger font-size, different typeface, etc.

Indentation of .njk files: The problem you're stating is that of code migration, a problem you'll have anyway, depending on where you're migrating from. If you copy-paste some code from a source file on the web, it's actually more likely to be indented with 2 spaces. We could do some research to see if it's possible to auto-indent the .njk files as well, but changing it back to 4 spaces is inconsistent.

Performance: The `codestyle` task I've added is not really meant for manually running it after each change, while it does work. It formats your entire project, and thus performance is accordingly. There are IDE integrations which format just the files you're editing, and I think your solution lies there. The current task can't really be optimized any further until Prettier introduces an asynchronous api.

---
@nienkedekker:
If you chose to run the task via npm (`npm run codestyle-is-valid`), npm will fail with that error message padded. But you don't have to, and running it via npm is mainly intended for the githooks (who need a certain exitcode). A human running this command would be better of running it via gulp: `gulp codestyle-is-valid` :)

I like your suggestion property order suggestion! And you're right, that should be a separate PR. It's a functional change., not a stylistic one. I agree with the property order, already using that personally.

**Comment by colin-aarts** [comment]
Wouldn't using tabs instead of spaces solve the indentation problem for everyone? 😉

**Comment by mdecorte** [comment]
@branneman I agree with @awolkers  
Since we are using prettier in conjunction with ESLint we should follow the (easy & minimalistic) recommended way: [docs](https://prettier.io/docs/en/eslint.html). Which is by using [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier).
The eslintconfig can be really small and there is no need for a prettierconfig file.

**Comment by peeke** [comment]
My 2 cents:

I really likes that this makes it easier for me to work on different projects. I work on a lot of different projects and have encountered quite some different linting configurations. I can get used to a codestyle different from my own, but when I have to switch so often I tend to forget a semicolon or something. Since my githooks usually don't work, this results in failing builds or colleagues getting my linting errors when they commit. Prettier fixes this. 

I don't really mind if anyone changes the chosen prettier settings, since there will still be a codestyle enforced throughout the project. Would be nice though.

I think the Nunjucks issue is a weird one, since Prettier shouldn't reformat that? (since it doesn't list support for it). I did see they have a plugin system in beta, which might mean a html/njk plugin in the future.

@MartijnNieuwenhuizen 
I don't think recognizing your code will really be an issue since the function names remain the same and it's impossible to keep a mental model of all of your code anyway. I mainly navigate my code using Cmd + F 😜  How do others navigate?

Indenting with 4 spaces or a tab might be a good idea if it helps your dislexia. Have you tried using 2 spaces to compare? And have you tried a larger line-height setting?

**Comment by MartijnNieuwenhuizen** [comment]
@peeke I still think an indentation will improve the readability because it's easier to spot on which line you are. But I'm not sure if it has anything to do with dyslexia. Therefore I tested the line-height, font-size and font-family like @branneman also suggested and that helps a lot!

The plugin system for the Nunjucks part is a nice one, but I could settle with the code reformat shortcut in Webstorm for the time being. 

Therefore, if the merge conflicts are fixed and the new commit is working correctly, you have my approval.

File: tasks/js.js [hunk 1]
```diff
@@ -1,104 +1,110 @@
...
+    config.js.src.components,
+    `!${config.js.src.vendor}`
+  ]
+
+  return gulp
+    .src(src)
+    .pipe(
+      eslint({
+        fix: config.js.eslintAutofix,
+        ignorePath: config.codestyle.lintignore
```
Comment (awolkers) [inline tasks/js.js]: This doesn't work. When I run `gulp js-lint` files within the vendor folder are also being linted.

Comment (awolkers) [reply tasks/js.js]: Both `codestyle.js` and `js.js` throw errors when I run the `gulp js-lint` task. Guess these need to be fixed as well because else these erros need to be fixed the first time someone commits a `.js` file when starting a new project.

**Comment by awolkers** [comment]
I did some more research on the comments about the eslint config file I did earlier. Extending the config is indeed a good way to go. I suggest adding the code below to the `.eslintrc` file just right after `parser: babel-eslint`.

```
extends:
  - eslint:recommended
  - prettier
``` 

In order to get this to work the following package needs to be added to the project: https://github.com/prettier/eslint-config-prettier.

Also all the rules within the `.eslintrc` file need to be checked again since lots of them will be redundant. In this list https://eslint.org/docs/rules/ rules that are defined within `eslint:recommended` are marked with a check mark.

File: tasks/js.js (hunk 1 above)
Comment (nienkedekker) [reply tasks/js.js]: What I think works is moving `ignorePath: config.codestyle.lintignore` from `tasks/js.js` to `config.js` as `ignorefile: '.lintignore'`:

```
js: {
		src: {...}
		eslintAutofix: false,
		ignorefile: '.lintignore'
	},
```

**Comment by peeke** [comment]
Options for prettier are defined in the package.json. Wouldn't it be better to use the .prettierrc file, as in their docs?

**Comment by branneman** [comment]
> Options for prettier are defined in the package.json. Wouldn't it be better to use the .prettierrc file, as in their docs?

@peeke: Because I wanted to skip adding **another** dotfile. If it were up to me, more setting files can go into the `package.json` since most of the tools we use support it anyway.

Prettier uses the same dependency for settings loading as the rest of them iirc, so it should be in their docs.

**Comment by peeke** [comment]
Ah, missed that. That's actually nice. 

I do have a question though. I've pasted my .prettierrc from another project and expected it to work. I didn't have the prettier field in my package.json yet. I would've expected it to detect the .prettierrc in that case. Do we have codestyle task hardcoded to look for options in the package.json?

**Comment by awolkers** [comment]
@peeke Yes, it's hardcoded within `codestyle.js`. I also expected it not to be.

File: package.json [hunk 2]
```diff
@@ -60,28 +67,30 @@
...
+    "ignore": "^3.3.7",
     "js-beautify": "^1.6.12",
     "js-yaml": "^3.6.1",
     "marked": "^0.3.6",
     "mkdirp": "^0.5.1",
     "moment-timezone": "^0.5.3",
+    "prettier": "1.9.2",
     "require-globify": "^1.4.1",
     "run-sequence": "^1.1.5",
-    "stylelint-config-standard": "^16.0.0",
```
Comment (awolkers) [inline package.json]: Please don't delete this but replace it with https://github.com/shannonmoeller/stylelint-config-prettier

File: .eslintrc.yml [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  no-dupe-args: error
+  no-dupe-keys: error
+  no-duplicate-case: error
+  no-empty:
+    - error
+    - allowEmptyCatch: true
+  no-empty-character-class: error
+  no-ex-assign: error
+  no-extra-boolean-cast: error
+  no-extra-parens: error
```
Comment (awolkers) [inline .eslintrc.yml]: This should be removed, this conflicts with prettier configuration

**Comment by awolkers** [comment]
I just updated this PR based on the comments I made earlier.

- Fixed some linting errors
- Fixed wrong config reference to ESlint ignorePath
- Added Prettier rulesets for both ESlint as Stylelint and updated existing ruleset (see below)

### ESlint 
We start with the recommended ruleset from ESlint itself. This ruleset is extended with a ruleset that is optimized for Prettier. Finally we add/override rules with our own config.

### Stylelint
We now start with a ruleset optimized for Prettier and then we add/override this set with our own config.

File: .editorconfig [hunk 4]
```diff
@@ -1,21 +1,9 @@
-# EditorConfig - for more info check http://EditorConfig.org
-
-# top-most EditorConfig file
 root = true
 
 [*]
 charset = utf-8
 end_of_line = lf
 indent_style = space
-indent_size = 4
```
[the remark this answers is not in the document]
Comment (awolkers) [reply .editorconfig]: Home.njk wasn't updated to the new indentation. I fixed this.

**Comment by MartijnNieuwenhuizen** [comment]
The `.babelrc` file isn't updated with the new indentation yet. Otherwise approved
